### PR TITLE
Transaction feature

### PR DESCRIPTION
**Pull request by Ketan-SP18**
- Add TransactionController, TransactionRepository, TransactionService with base code. Will do specific implementation in JPG-006, JPG-007
- Add TransactionEntity, TransactionRequest, TransactionResponse, TransactionType
- Add TODO for specific implementation

### CODE REVIEW

File: src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import javax.validation.constraints.NotNull;
+import java.time.LocalDateTime;
+
+@Data
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class TransactionResponse {
+
+    private Float amount;
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java:21]: What if i want to see a specific  transaction ?
In response there is no transaction id?

File: src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    @Column(name = "txt_date_time")
+    @NotNull
+    @JsonFormat(pattern = "MMM dd, yyyy hh:mm:ss a", timezone = "Asia/Kolkata")
+    private LocalDateTime time;
+
+    @GeneratedValue(generator = "UUID")
+    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
+    @Column(name = "ref_number")
+    private UUID referenceNumber;
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java:58]: What is this for can you explain the significance?

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        this.userRepository = userRepository;
+    }
+
+    public TransactionResponse transact(Integer userID, TransactionRequest request) throws Exception {
+
+        Optional<UserEntity> optionalUser = userRepository.findById(userID);
+        if (optionalUser.isEmpty())
+            return TransactionResponse.builder().message("User not found. Please use a registered user").build();
+
+//        UserEntity user = optionalUser.get();
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:31]: Remove this line

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+        // TODO: Add deposit/withdraw logic and change below return accordingly
+
+        return TransactionResponse.builder().build();
+    }
+
+    private Float getZeroOrLastBalance(UserEntity user) {
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:38]: Remove this method too.

**Review by spark-coders — COMMENTED** [review]
Java docs are still missing.
Add them and push the code ASAP.

File: src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java (hunk 2 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java:58]: It is reference number to identify particular transaction

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 4 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:38]: Currently required to show closing balance

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 3 above)
Comment (Bhavesh-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:31]: yes sir

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.validation.Valid;
+
+/**
+ * The type Transaction controller.
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java:18]: Update the java doc properly

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+     * Instantiates a new Transaction controller.
+     *
+     * @param transactionService the transaction service
+     */
+    public TransactionController(TransactionService transactionService) {
+        this.transactionService = transactionService;
+    }
+
+    /**
+     * Transact response entity.
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java:35]: Update the java doc properly

File: src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java (hunk 1 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java:26]: Added transaction id

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java (hunk 5 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java]: Done

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java (hunk 6 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java]: done

File: src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionRequest.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import javax.persistence.EnumType;
+import javax.persistence.Enumerated;
+import javax.validation.constraints.NegativeOrZero;
+import javax.validation.constraints.NotNull;
+
+/**
+ * details/fields of transaction to be requested
```
[the remark this answers is not in the document]
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionRequest.java]: done

File: src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import lombok.NoArgsConstructor;
+
+import javax.persistence.EnumType;
+import javax.persistence.Enumerated;
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.NotNull;
+import java.time.LocalDateTime;
+
+/**
+ * details/fields of the transaction requested are displayed
```
[the remark this answers is not in the document]
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java]: done

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+     * @param transactionRepository to perform operation related to transaction repository like finding user from transaction repository
+     * @param userRepository to perform operation related to user repository like finding user id from user repository
+     */
+    public TransactionService(TransactionRepository transactionRepository, UserRepository userRepository) {
+        this.transactionRepository = transactionRepository;
+        this.userRepository = userRepository;
+    }
+
+    /**
+     * Transact transaction response.
```
[the remark this answers is not in the document]
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java]: done

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * Instantiates a new Transaction controller.
+     * @param transactionService the transaction service is instantiated to perform further operation
+     */
+    public TransactionController(TransactionService transactionService) {
+        this.transactionService = transactionService;
+    }
+
+    /**
+     * Transact response entity gives the transaction response
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java:34]: Not appropriate java doc

File: src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * Transaction response where in response is generated on performing a transaction.
+ */
+@Data
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class TransactionResponse {
+
+    private Integer txnID;
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java:24]: Java doc foe fields?

File: src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java (hunk 2 above)
Comment (spark-coders) [reply src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java:63]: Why still  no replay no action?

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.util.Optional;
+
+/**
+ * Transaction service helps to provide service to perform transactions
+ */
+@Service
+public class TransactionService {
+
+    private final TransactionRepository transactionRepository;
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:18]: java docs for fields

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        if (optionalUser.isEmpty())
+            return TransactionResponse.builder().message("User not found. Please use a registered user").build();
+
+        // TODO: Add deposit/withdraw logic and change below return accordingly
+
+        return TransactionResponse.builder().build();
+    }
+
+    /**
+     * This method(getZeroOrLastBalance) is basically used for showing user closingBalance.
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:50]: Can be better refined

**Review by spark-coders — CHANGES_REQUESTED** [review]

File: src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java (hunk 2 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java:63]: removed

File: src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionRequest.java [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+ */
+@Data
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class TransactionRequest {
+
+    @NotNull(message = "Transaction amount cannot be null")
```
[the remark this answers is not in the document]
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionRequest.java:22]: Done

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java (hunk 10 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java:34]: updated

File: src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java (hunk 11 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java:24]: added

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 12 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:18]: added

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 13 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:50]: Updated

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * To process incoming REST API requests for Transactions and give response from the request body and connecting through transaction service (business logic) is carried out here with the help of Transaction Controller
+ */
+@RequestMapping("/api/v1/user")
+@RestController
+public class TransactionController {
+    private final TransactionService transactionService;
+
+    /**
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java:25]: this one not needed

File: src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    @JsonFormat(pattern = "MMM dd, yyyy hh:mm:ss a", timezone = "Asia/Kolkata")
+    private LocalDateTime time;
+
+    @Column(name = "type")
+    @Enumerated(EnumType.STRING)
+    private TransactionType transactionType;
+
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "user_id", nullable = false)
+    @OnDelete(action = OnDeleteAction.CASCADE)
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java:64]: Nice

File: src/main/java/com/passbook/sparkeighteen/peristence/repository/TransactionRepository.java [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+ */
+@Repository
+public interface TransactionRepository extends JpaRepository<TransactionEntity, Integer> {
+
+    /**
+     * Find by user optional.
+     * @param user function to perform on this particular user
+     * @return the optional - return th user or null
+     */
+    Optional<List<TransactionEntity>> findByUser(UserEntity user);
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/peristence/repository/TransactionRepository.java:22]: Don't wrap any collection into optional

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+    /**
+     * transactionRepository field is used to perform operation related to searching in database of transaction repository
+     */
+    private final TransactionRepository transactionRepository;
+    /**
+     * userrepository field is used to perform operation related to search in database of user repository
+     */
+    private final UserRepository userRepository;
+
+    /**
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:27]: not needed

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+ * Transaction service helps to provide service to perform transactions
+ */
+@Service
+public class TransactionService {
+
+    /**
+     * transactionRepository field is used to perform operation related to searching in database of transaction repository
+     */
+    private final TransactionRepository transactionRepository;
+    /**
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:22]: not needed

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.Optional;
+
+/**
+ * Transaction service helps to provide service to perform transactions
+ */
+@Service
+public class TransactionService {
+
+    /**
+     * transactionRepository field is used to perform operation related to searching in database of transaction repository
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:19]: not needed

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+     * @return the transaction response. (CREDIT OR DEBIT)
+     * @throws Exception shows error messages.
+     */
+    public TransactionResponse transact(Integer userID, TransactionRequest request) throws Exception {
+
+        Optional<UserEntity> optionalUser = userRepository.findById(userID);
+        if (optionalUser.isEmpty())
+            return TransactionResponse.builder().message("User not found. Please use a registered user").build();
+
+        // TODO: Add deposit/withdraw logic and change below return accordingly
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:50]: where is this logic?

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+
+    /**
+     * This method(getZeroOrLastBalance) is to get balance as zero if its user first transaction or the closing balance of the latest transaction.
+     * first assign the balance 0.
+     * @param user to get the balance of that specific user.
+     * @return updated balance after transaction done.
+     */
+    private Float getZeroOrLastBalance(UserEntity user) {
+        Float balance = 0f;
+        // TODO: Add getting the latest closing balance of the user
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:63]: logic?

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+        return TransactionResponse.builder().build();
+    }
+
+    /**
+     * This method(getZeroOrLastBalance) is to get balance as zero if its user first transaction or the closing balance of the latest transaction.
+     * first assign the balance 0.
+     * @param user to get the balance of that specific user.
+     * @return updated balance after transaction done.
+     */
+    private Float getZeroOrLastBalance(UserEntity user) {
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:61]: getUpdatedBalance

**Review by kunalcodess — CHANGES_REQUESTED** [review]

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 21 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:50]: This is just  the base transaction branch

File: src/main/java/com/passbook/sparkeighteen/peristence/repository/TransactionRepository.java (hunk 17 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/repository/TransactionRepository.java:22]: yes will keep in mind

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 23 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:61]: yes will refactor it

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 20 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:19]: needed further

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 19 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:22]: needed later

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java (hunk 15 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java:25]: required further

File: src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionRequest.java [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * Transaction request where in request is initiated to perform transaction
+ */
+@Data
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class TransactionRequest {
+
+    /**
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionRequest.java:22]: not needed

File: src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * Transaction response where in response is generated on performing a transaction.
+ */
+@Data
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class TransactionResponse {
+
+    /**
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java:24]: not needed

**Review by kunalcodess — CHANGES_REQUESTED** [review]

**Pull request merged**